### ISSUE TITLE
Potential Include Bug

### ISSUE THREAD

**Issue by RKennedy9064** [json-api-dotnet/JsonApiDotNetCore#381]
## Description

I've been working on a workaround for sorting using nested relationships and have run into something weird. I've added an attribute on my model to expose the nested attribute like so.

```c#
[NotMapped]
[Attr("staff-roster-last-name")]
public string StaffRosterLastName => 
  StaffRoster != null ? StaffRoster.LastName : String.Empty;
```

I'm able to access the attribute and display it on the front end, but when I try to sort by staff-roster-first-name, StaffRoster is always null when the sort filters are applied. The weird thing that I've noticed during various tests is that if I add a call ToList on the entities inside of EntityResourceService.cs GetAsync(), StaffRoster becomes available and the sorting works correctly. I know calling ToList executes the query, I'm just not sure why the included model isn't available unless I add the line of code. My url looks like this.

```
http://localtest.me:5000/api/v1/sample-units?include=field-disposition%2Csample-disposition%2Cstaff-roster&page%5Bnumber%5D=1&page%5Bsize%5D=25&sort=staff-roster-last-name
```

And the code I'm referring to looks like so.

```c#
public virtual async Task<IEnumerable<TResource>> GetAsync()
{
  var entities = _entities.Get();

  entities = ApplySortAndFilterQuery(entities);

  if (ShouldIncludeRelationships())
    entities = IncludeRelationships(entities, _jsonApiContext.QuerySet.IncludedRelationships);

  if (_jsonApiContext.Options.IncludeTotalRecordCount)
    _jsonApiContext.PageManager.TotalRecords = await _entities.CountAsync(entities);

  // This is the line I added.
  entities.ToList();

  // pagination should be done last since it will execute the query
  var pagedEntities = await ApplyPageQueryAsync(entities);
  return pagedEntities;
}
```

For some reason the only way I can get StaffRoster to not be null and sort correctly is if I explicitly call entities.ToList() before ApplyPageQueryAsync is called. Looking through the paging code it looks like eventually ToListAsync() is called. so I'm not sure why an extra call of ToList is needed for the sort to work correctly. Do you have any idea why this might be the case? Any input would be greatly appreciated.

## Environment

- JsonApiDotNetCore Version: Master branch as of 8/17/2018
- Other Relevant Package Versions: N/A

**Comment by RKennedy9064** [comment]
@jaredcnance I was looking into this over the weekend and it looks like it might have to do with the fact that my NotMapped property isn't available until the query is materialized with ToList(), which is why it works when adding it. Do you have any suggested workarounds that might allow me to sort by a related field until that feature is implemented?

**Comment by jaredcnance** [comment]
Yeah that was my initial thought as well, but it's a little unclear why sorting works at all since in your example there is nothing that should apply a sort AFTER materializing the query results. 

For now, this is what I would recommend creating an override for `ApplySortAndFilterQuery` that explicitly handles the included field. Something like:

```c#
protected override IQueryable<TEntity> ApplySortAndFilterQuery(IQueryable<TEntity> entities)
{
    // ...
    foreach (var filter in query.Filters) {
        entities = (filter.Key == "staff-roster-last-name")
            ? CustomNestedIncludeSort(entities, filter) // < ---
            : _entities.Filter(entities, filter);
    }
    // ...
}
```

Where the default implementation looks like:

https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/23329ab1b6d20494e29e7943fec1acea8bd7ca6a/src/JsonApiDotNetCore/Services/EntityResourceService.cs#L215-L229

**Comment by RKennedy9064** [comment]
Yeah I'm a little confused as to why it worked after materializing also, especially since it was materialized after the sort had been applied. I'll just override the ApplySortAndFilterQuery to handle the specific cases I need. Thanks for the help.

**Comment by RKennedy9064** [comment]
@jaredcnance I'm currently working on implementing the sorting and had a quick question. The example you posted is looping through the filters, but since I'm sorting, will I need to modify the sort extension method. It looks like that's where everything gets sent to when applying the sorts.

```
public static IQueryable<TSource> Sort<TSource>(this IQueryable<TSource> source, List<SortQuery> sortQueries)
{
    if (sortQueries == null || sortQueries.Count == 0)
        return source;

    var orderedEntities = source.Sort(sortQueries[0])
            if (sortQueries.Count <= 1) return orderedEntities;
                for (var i = 1; i < sortQueries.Count; i++)
                    orderedEntities = orderedEntities.Sort(sortQueries[i]);

    return orderedEntities;
}
```

**Comment by jaredcnance** [comment]
@RKennedy9064 since nested filters and sorting is available in 3.0.0, can this be closed?

**Comment by jaredcnance** [comment]
Awesome! Actually the credit for that goes to @milosloub!
